core/src/main/java/org/apache/iceberg/BaseManifestListFile.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import org.apache.iceberg.encryption.Ciphers;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.encryption.EncryptionUtil;
+import org.apache.iceberg.encryption.KeyEncryptionKey;
+import org.apache.iceberg.encryption.StandardEncryptionManager;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ByteBuffers;
+
+public class BaseManifestListFile implements ManifestListFile, Serializable {
+  private final String location;
+  private final String mdEncryptionKeyID;
+  private final ByteBuffer encryptedKeyMetadata;
+  private ByteBuffer keyMetadata;
+
+  public BaseManifestListFile(
+      String location,
+      ByteBuffer keyMetadata,
+      String mdEncryptionKeyID,
+      ByteBuffer encryptedKeyMetadata) {
+    this.location = location;
+    this.keyMetadata = keyMetadata;
+    this.encryptedKeyMetadata = encryptedKeyMetadata;
+    this.mdEncryptionKeyID = mdEncryptionKeyID;
+  }
+
+  @Override
+  public String location() {
+    return location;
+  }
+
+  @Override
+  public ByteBuffer keyMetadata() {
+    return keyMetadata;
+  }
+
+  @Override
+  public String metadataEncryptionKeyID() {
+    return mdEncryptionKeyID;
+  }
+
+  @Override
+  public ByteBuffer encryptedKeyMetadata() {
+    return encryptedKeyMetadata;
+  }
+
+  public void setDecryptedKeyMetadata(ByteBuffer decryptedKeyMetadata) {
+    this.keyMetadata = decryptedKeyMetadata;
+  }
+
+  static ManifestListFile create(
+      String location, EncryptionManager em, EncryptionKeyMetadata keyMetadata, long length) {
```

Comment:
I think this is a method that should live in `EncryptionUtil` rather than here. This also needs the current key encryption key to be passed in.

Reply:
> this is a method that should live in EncryptionUtil rather than here

SGTM

> This also needs the current key encryption key to be passed in.

Per the other comments, the current KEK is the latest entry in the KEK cache, kept in the StandardEncryptionManager (accessible by the EncryptionUtil, but not by the SnapshotProducer that calls this "create manifest list" method).